sanitycheck: Add fixture option for external hardware dependency

pswarnak opened the pull request:
Add a new option as fixture in harness configurations for utilizing
sanitycheck to identify test cases that require external hardware
such as sensor, ble, networking for validation. The config will be
added to yaml files with unique fixture name to identify each hardware
and allow automation to trigger test execution on setup having the
specific fixture enabled.

Signed-off-by: Praful Swarnakar <praful.swarnakar@intel.com>

inakypg reviewed (commented):
LGTM, but I'll let @nashif  comment on the final approval.

pswarnak committed:
sanitycheck: Add fixture option for external hardware dependecy

Add a new option as fixture in harness configurations for utilizing
sanitycheck to identify test cases that require external hardware
such as sensor, ble, networking for validation. The config will be
added to yaml files with unique fixture name to identify each hardware
and allow automation to trigger test execution on setup having the
specific fixture enabled. Also, remove the default required for type and
regex configs that is not essential in case of ztest based test cases.

Signed-off-by: Praful Swarnakar <praful.swarnakar@intel.com>

erwango commented on samples/sensor/hts221/sample.yaml:
```diff
@@ -1,7 +1,17 @@
+      regex:
+        - "Temperature:(.*)C"
+        - "Relative Humidity:(.*)%"
+      fixture: fixture_i2c_hts221 # see documentation
```
Can you provide an example on how this should be populated on board side?
For instance, on disco_l475_iot1 that embeds hts221 sensor

pswarnak replied:
@erwango:
> Can you provide an example on how this should be populated on board side?

This entirely depends on the way you use this information. Here, the fixture "fixture_i2c_hts221" is just a way of saying that this test case that is dependent on an external device(HTS221 sensor in this case) and it should be pick a device that fulfills this dependency. So, it entirely depends on your test setup on how you design the board selection(pick one board out of multiple in an automation setup) logic based on "fixture" keyboard.

> For instance, on disco_l475_iot1 that embeds hts221 sensor

For the specific board disco_l475_iot1 that embeds hts221 sensor, you can just have another case specified in sample.yaml such as:
```
  sensor.hts221.embedded:
    tags: sensor
    # disco_l475_iot1 has hts221 sensor present internally
    platform_whitelist: disco_l475_iot1
    depends_on: i2c
```

nashif commented on samples/sensor/hts221/sample.yaml:
```diff
@@ -1,7 +1,18 @@
+      regex:
+        - "Temperature:(.*)C"
+        - "Relative Humidity:(.*)%"
+      fixture: fixture_i2c_hts221 # see documentation
```
> see documentation.. 

what documentation?

inakypg replied:
still please remember our voice convo, where we said this TC shall be modified not to do an expect sequence, as it is actually a ztest -- if a condition happens which would make the _printing_ impossible, then it shall raise a failure/error

pswarnak replied:
@nashif: Added documentation with my limited knowledge about harness and it's various config options. Additionally, mentioned how to and where to use fixture keyword. Kindly check.

pswarnak replied:
@inakypg: The current test that is posted here is part of "samples" that are not ztest based. So, I think regex is required as the test will not print any "PASS" or "PRIJECT EXECUTION SUCCESSFUL" statement.
However, we will not be using regex option for ztest based test cases as suggested.

nashif commented on samples/sensor/hts221/sample.yaml:
```diff
@@ -1,7 +1,18 @@
     tags: samples sensor
-    depends_on: i2c hts221
+    depends_on: i2c
+    platform_whitelist: quark_d2000_crb
```
why is this being whitelisted on quark_d2000? Having to remove the depend_on on hts221 is going to cause problems, I know why this was added and was a workaround, but it seems we have a conflict between the fixture and depends_on.

pswarnak replied:
> why is this being whitelisted on quark_d2000?

Sanitycheck reports build failure for this test(samples/sensor/hts221) on multiple boards that have i2c capability, but does not implement hts221 sensor driver. I have validated it on quark_d2000 that is connected in our automation. So, whitelisting only on particular platform that passes the test to avoid sanity check failure.

**Sanity check failure Log:**
```
[ 89%] Building C object zephyr/drivers/sensor/hts221/CMakeFiles/drivers__sensor__hts221.dir/hts221.c.obj
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/sensor/hts221/hts221.c: In function ‘hts221_init’:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/sensor/hts221/hts221.c:110:37: error: ‘CONFIG_HTS221_I2C_MASTER_DEV_NAME’ undeclared (first use in this function)
  drv_data->i2c = device_get_binding(CONFIG_HTS221_I2C_MASTER_DEV_NAME);
                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
**Shippable link:** https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/22154/3/tests
The following boards report this error: frdm_kl25z hexiwear_k64 frdm_k64f olimexino_stm32 nucleo_f070rb stm32f072b_disco stm32f3_disco frdm_kw41z reel_board nucleo_l053r8 96b_carbon stm32f723e_disco nucleo_f429zi

@inakypg: Can you please suggest a better way instead for hard-coding for a particular platform as platform_whitelist. Possibly, one way is to enable hts221 dependency only for the boards that support this sensor driver. Currently, hts221 is just a dummy value here as rightly mentioned by @nashif.

pswarnak committed:
doc: Add description for harness and its various configs

Document how to use harness and it's various configurations
such as type, regex and fixture that can be used to pick specific
board setup that fulfills external dependency for test case.

Signed-off-by: Praful Swarnakar <praful.swarnakar@intel.com>

dbkinder commented on doc/subsystems/test/sanitycheck.rst:
```diff
@@ -385,6 +385,64 @@ extra_sections: <list of extra binary sections>
     extra, unexpected sections in the Zephyr binary unless they are named
     here. They will not be included in the size calculation.
 
+harness: <list of harness>
```
plural of harness is harnesses, so this should be ``list of harnesses``

pswarnak replied:
Changed as suggested by @nashif.

dbkinder commented on doc/subsystems/test/sanitycheck.rst:
```diff
@@ -385,6 +385,64 @@ extra_sections: <list of extra binary sections>
+    special harness is needed to run the tests sucessfully. This can be
+    as simple as a loopback wiring or a complete hardware test setup for
+    sensor and IO testing. Usually pertains to external dependency domains
+    but can be anything such as console, sensor, net, keyboard, bluetooth.
```
change to:  ``keyboard, or Bluetooth.``  (add the ``or`` and capitalize ``Bluetooth`` since it's a trademark)

pswarnak replied:
Thanks for figuring it out. Addressed.

dbkinder commented on doc/subsystems/test/sanitycheck.rst:
```diff
@@ -385,6 +385,64 @@ extra_sections: <list of extra binary sections>
+harness_config: <harness configuration options>
+    Extra harness configuration options to be used to select a board and/or
+    for handling generic Console with regex matching. Config can announce
+    what features it supports, this option will enable the test to run on
```
change comma to a period and start a new sentence: ``it supports.  This option``

pswarnak replied:
Done

dbkinder commented on doc/subsystems/test/sanitycheck.rst:
```diff
@@ -385,6 +385,64 @@ extra_sections: <list of extra binary sections>
+        Number of times to validate the repeated regex expression
+
+    fixture: <expression>
+        It is a way of specifying test case dependency on an external
```
Change to:

    Specify a test case dependency on an external device
    (e.g., sensor), and identify boards that fulfill this dependency.

pswarnak replied:
Done

dbkinder commented on doc/subsystems/test/sanitycheck.rst:
```diff
@@ -385,6 +385,64 @@ extra_sections: <list of extra binary sections>
+        It is a way of specifying test case dependency on an external
+        device(E.g. Sensor) and it can be used to identify board that fulfills
+        this dependency. It depends on specific test setup and board selection
+        logic to pick the particular board(s) out of multiple boards that fulfil
```
s/fulfil/fulfill/

pswarnak replied:
Addressed. Kindly check

dbkinder commented on doc/subsystems/test/sanitycheck.rst:
```diff
@@ -385,6 +385,64 @@ extra_sections: <list of extra binary sections>
     here. They will not be included in the size calculation.
 
+harness: <list of harness>
+    A set of string harness for the testcase. This keyword means that a
```
Not sure what "a set of string harness" is... perhaps merge these two sentences into:

    A list of harnesses needed to run the tests successfully.

pswarnak replied:
Addressed. Kindly check.

nashif commented on doc/subsystems/test/sanitycheck.rst:
```diff
@@ -385,6 +385,63 @@ extra_sections: <list of extra binary sections>
     extra, unexpected sections in the Zephyr binary unless they are named
     here. They will not be included in the size calculation.
 
+harness: <list of harnesses>
```
this is not a list.

pswarnak replied:
Fixed. Thanks for figuring out.

nashif commented on doc/subsystems/test/sanitycheck.rst:
```diff
@@ -385,6 +385,63 @@ extra_sections: <list of extra binary sections>
+
+    fixture: <expression>
+        Specify a test case dependency on an external device(e.g., sensor),
+        and identify boards that fulfills this dependency. It depends on
```
and identify setups that fulfil this dependency...

pswarnak replied:
Modified

pswarnak commented:
@dbkinder:
Thank you for thorough review. Addressed all review comments. Kindly check.

pswarnak commented:
@nashif:
Modified as per suggestion, removed the sensor sample file and added gpio loopback yaml for better clarity. Kindly review.
Also, created new file with current defined fixture list. Please confirm, if the data and file format is fine Or looking for any specific format.

dbkinder commented on doc/subsystems/test/sanitycheck.rst:
```diff
@@ -385,6 +385,63 @@ extra_sections: <list of extra binary sections>
+
+    fixture: <expression>
+        Specify a test case dependency on an external device(e.g., sensor),
+        and identify setups that fulfills this dependency. It depends on
```
so close.  change ``fulfills`` to ``fulfill``

pswarnak replied:
Sorry, missed it. Done now.

dbkinder reviewed (dismissed):
one typo

pswarnak commented:
@nashif: Modified as per suggestion. Kindly review. Thank you.
